theano/gpuarray/c_code/dnn_fwd.c

```diff
@@ -72,24 +76,32 @@ APPLY_SPECIFIC(conv_fwd)(PyGpuArrayObject *input, PyGpuArrayObject *kerns,
     }
     return 0;
   }
-
-  if (c_set_tensor_for_conv(input, APPLY_SPECIFIC(input), params->num_groups) == -1)
+  
+  int groups = c_set_groups_for_conv(desc, params->num_groups);
+  if (groups == -1)
+    return 1;
+  if (c_set_tensor_for_conv(input, APPLY_SPECIFIC(input), groups) == -1)
     return 1;
-  if (c_set_filter(kerns, APPLY_SPECIFIC(kerns), params->num_groups) == -1)
+  if (c_set_filter(kerns, APPLY_SPECIFIC(kerns), groups) == -1)
     return 1;
-  if (c_set_tensor_for_conv(*output, APPLY_SPECIFIC(output), params->num_groups) == -1)
+  if (c_set_tensor_for_conv(*output, APPLY_SPECIFIC(output), groups) == -1)
     return 1;
-  size_t input_offset = PyGpuArray_STRIDE(input, 0) / params->num_groups;
-  size_t kern_offset = PyGpuArray_STRIDE(kerns, 0) * PyGpuArray_DIM(kerns, 0) / params->num_groups;
-  size_t output_offset = PyGpuArray_STRIDE(*output, 0) / params->num_groups;
+  size_t input_offset = PyGpuArray_STRIDE(input, 0) / groups;
+  size_t kern_offset = PyGpuArray_STRIDE(kerns, 0) * PyGpuArray_DIM(kerns, 0) / groups;
+  size_t output_offset = PyGpuArray_STRIDE(*output, 0) / groups;
 
   cudnnConvolutionFwdAlgo_t algo = params->conv_algo;
+  size_t   worksize  = 0;
+  cudnnMathType_t mathtype = CUDNN_DEFAULT_MATH;
+  
+  std::string hashkey;
+  
   #ifdef DEBUG
   char algorithm_name[128];
   #endif
 
   cuda_enter(c->ctx);
-
+  
   if (params->choose_algo) {
```

Comment:
IRL discussion with @notoraptor. We think we should get rid of the prev_img_dims and prev_kern_dims (and this whole if).

This seem to allow to bypass your cache if the input and kernel shapes don't change. In that case, it set reuse_algo to 1. This won't look at the other characteristic your cache look at.

Do you think we should care the the speed to compute the hash and query the cache? I have the impression we could also query it before each convolution, but being sure of that would be great.